### PR TITLE
Document scala.Singleton

### PR DESCRIPTION
**Pull request by callumcodes**
Information from: http://scala-programming-language.1934581.n4.nabble.com/scala-Singleton-td1940630.html

Fixes https://github.com/scala/bug/issues/9525

### CODE REVIEW

File: src/library-aux/scala/Singleton.scala [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+/** `Singleton` is not usually used directly. It is essentially an encoding trick for existentials.
+  *
+  *  {{{
+  *  T forSome { val x: T }
+  *  }}}
+  *  is turned into
+  *  {{{
+  *  [x.type := X] T forSome { type X <: T with Singleton }
+  *  }}}
```
Comment (lrytz) [inline src/library-aux/scala/Singleton.scala]: I think is a little too technical for a Scaladoc comment. Also, `Singleton` has some valid use cases in user code. My suggestion:

```
`Singleton` is used by the compiler as a supertype for singleton types. This includes literal types,
as they are also singleton types.

{{{
scala> object A { val x = 42 }
defined object A

scala> implicitly[A.type <:< Singleton]
res12: A.type <:< Singleton = generalized constraint

scala> implicitly[A.x.type <:< Singleton]
res13: A.x.type <:< Singleton = generalized constraint

scala> implicitly[42 <:< Singleton]
res14: 42 <:< Singleton = generalized constraint

scala> implicitly[Int <:< Singleton]
                 ^
       error: Cannot prove that Int <:< Singleton.
}}}

`Singleton` has a special meaning when it appears as an upper bound on a formal type
parameter. Normally, type inference in Scala widens singleton types to the underlying
non-singleton type. When a type parameter has an explicit upper bound of `Singleton`,
the compiler infers a singleton type.

{{{
scala> def check42[T](x: T)(implicit ev: T =:= 42): T = x
check42: [T](x: T)(implicit ev: T =:= 42)T

scala> val x1 = check42(42)
                       ^
       error: Cannot prove that Int =:= 42.

scala> def singleCheck42[T <: Singleton](x: T)(implicit ev: T =:= 42): T = x
singleCheck42: [T <: Singleton](x: T)(implicit ev: T =:= 42)T

scala> val x2 = singleCheck42(42)
x2: Int = 42
}}}

See also [[https://docs.scala-lang.org/sips/42.type.html SIP-23 about Literal-based Singleton Types]].
```

**Review by lrytz — CHANGES_REQUESTED** [review]

**Comment by callumcodes** [comment]
@lrytz I have updated my PR to include your suggestion

**Review by lrytz — APPROVED** [review]

**Comment by lrytz** [comment]
Thanks!

**Comment by SethTisue** [comment]
thank you, and welcome!

**Pull request merged**